### PR TITLE
JAMES-3775 Write webadmin route to create feeding ham messages task

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

**Review by Arsnael — APPROVED** [review]

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedHamToRSpamDTask.java [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+        this.rSpamDHttpClient = rSpamDHttpClient;
+        this.context = new Context(runningOptions);
+        this.clock = clock;
+    }
+
+    @Override
+    public Result run() {
+        Optional<Date> afterDate = runningOptions.period.map(periodInSecond -> Date.from(clock.instant().minusSeconds(periodInSecond)));
+        try {
+            return messagesService.getMailboxMessagesOfAllUser(INBOX_MAILBOX_NAME)
```
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedHamToRSpamDTask.java:344]: Nope this should IMO be all messages except those in spam / trash folders.

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedHamToRSpamDTask.java [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    }
+
+    @VisibleForTesting
+    public Context.Snapshot snapshot() {
+        return context.snapshot();
+    }
+
+    public static boolean randomBooleanWithProbability(double probability) {
+        return Math.random() < probability;
+    }
```
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedHamToRSpamDTask.java:388]: Idem we do not need random if probability is 1.0

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedHamToRSpamDTask.java [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        try {
+            return messagesService.getMailboxMessagesOfAllUser(INBOX_MAILBOX_NAME)
+                .doOnNext(messageResult -> context.incrementHamMessageCount())
+                .filter(messageResult -> afterDate.map(date -> messageResult.getInternalDate().after(date)).orElse(true))
+                .filter(any -> randomBooleanWithProbability(runningOptions.samplingProbability))
+                .transform(ReactorUtils.<MessageResult, Result>throttle()
+                    .elements(runningOptions.messagesPerSecond)
+                    .per(Duration.ofSeconds(1))
+                    .forOperation(messageResult1 -> Mono.just(messageResult1)
+                        .flatMap(Throwing.function(messageResult -> rSpamDHttpClient.reportAsHam(messageResult.getHeaders().getInputStream(), messageResult.getBody().getInputStream())))
```
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/task/FeedHamToRSpamDTask.java:352]: Idem simplify this.

**Review by chibenwa — COMMENTED** [review]
Unit for period also.

**Comment by quantranhong1999** [comment]
2 latest fixup commits address comments of this PR and similar ones in Spam PR. The rest is basically the same.

**Comment by vttranlina** [comment]
The ci fail is not related to this pr

**Comment by vttranlina** [comment]
Can you fix the typo "RSpamD" to "Rspamd"?

**Comment by quantranhong1999** [comment]
> Can you fix the typo "RSpamD" to "Rspamd"?

Me what I don't like is that instead of 1 PR doing 1 job that fixes the typo, now what you suggest is others n PRs doing both their job plus the fix the typo stuff.

Me I would prefer the typo PR to do its clean job eventually.

**Pull request merged**